kernel/binding.cc

```diff
@@ -24,6 +24,163 @@ YOSYS_NAMESPACE_BEGIN
 RTLIL::Binding::Binding(RTLIL::IdString target_type,
                         RTLIL::IdString target_name)
 	: target_type(target_type), target_name(target_name)
-{}
+{
+	// A counter of the number of bindings that have been created so far.
+	static int binding_counter;
+	attr_name = stringf("\\bound%d", binding_counter++);
+}
+
+// Return true if mod is either called name or is derived from an abstract
+// module that was called name.
+static bool
+module_original_type_matches(const RTLIL::Module &mod, RTLIL::IdString name)
+{
+	if (mod.name == name) return true;
+
+	// A module derived from 'foo' will have a name starting with $paramod
+	// and an 'hdlname' attribute equal to '\foo'.
+	if (!mod.name.begins_with("$paramod"))
+		return false;
+
+	RTLIL::IdString hdlname = mod.get_string_attribute(ID::hdlname);
+	log_assert(!hdlname.empty());
+	return (hdlname == name);
+}
+
+// Search through the design for modules called name, or that are derived from
+// one called name.
+static std::vector<RTLIL::Module *>
+find_concrete_modules(RTLIL::Design &design, RTLIL::IdString name)
+{
+	std::vector<RTLIL::Module *> mods;
+
+	for (const auto &pr : design.modules_) {
+		if (module_original_type_matches(*pr.second, name))
+			mods.push_back(pr.second);
+	}
+
+	return mods;
+}
+
+// Try to find a cell below mod whose (dot-separated) name is given by path.
+// Modify so_far in place and return true on success.
+static bool
+find_cell_in(RTLIL::Binding::Path *so_far,
+             RTLIL::Design &design, RTLIL::Module &mod, const char *path)
+{
+	const char *dot = strchr(path, '.');
+	std::string cell_name =
+		dot ? std::string(path, dot - path) : std::string(path);
+
+	RTLIL::Cell *cell = mod.cell('\\' + cell_name);
+	if (!cell)
+		return false;
+
+	so_far->push_back(RTLIL::Binding::Item(&mod, cell));
+	if (!dot)
+		return true;
+
+	// We found a child cell, but there's more searching to do. Recurse to
+	// search the cell's module.
+	return find_cell_in(so_far, design, *cell->module, dot + 1);
+}
+
+RTLIL::Binding::Path
+RTLIL::Binding::find_rel_cell(RTLIL::Design &design, RTLIL::Module &start) const
+{
+	Path ret;
+	bool found = find_cell_in(&ret, design, start, target_name.c_str() + 1);
+	if (!found)
+		return Path();
+
+	log_assert(ret.size());
+	check_cell_type(design, *ret.back().second);
+	return ret;
+}
+
+std::vector<RTLIL::Binding::Path>
+RTLIL::Binding::find_tl_cells(RTLIL::Design &design) const
+{
+	std::vector<Path> paths;
+
+	std::string tgt_name = target_name.str();
+	size_t tgt_off = 0;
+
+	// The target name should always start with a backslash
+	log_assert(tgt_name[0] == '\\');
+
+	// Is the name absolute (starts with $root)? If so, just skip past that:
+	// we're at the top already.
+	if (tgt_name.compare(0, 7, "\\$root.") == 0)
```

Comment:
Ah, I think it *is* imposed, because of which function we're in. The `find_tl_cells` method is supposed to split something like `A.B.C` into "the cell at path B.C within module A". That's what the stuff to compute `mod_name` and `cell_path` does just below. To interpret `A.B.C` as "the cell at path A.B.C relative to the current point" needs the `find_rel_cell` method. There are some comments explaining what's supposed to happen in `binding.h`.

Or have I misunderstood your message?

Reply:
For example, the following is currently accepted, but shouldn't be, as `act` does not lie at the root of the instantiation tree.
```SystemVerilog
module foo;
  initial $display("hello from foo");
endmodule
module bar;
  initial $display("hello from bar");
endmodule
module act;
  initial $display("hello from act");
  foo f();
endmodule
module beep;
  act act1();
endmodule
module boop;
  beep b1();
endmodule
bind $root.act.f bar b();
```

Reply:
You're completely right! I hadn't been distinguishing between "top-level, so a module name" and "the top of the instantiation tree". Will fix.